File: dev/benchmark.py
import time
from dt_core_python import *

consumer = DTLogConsumer("log", 200, "dt_py", 10*1024*1024)
dt = DTAnalytics(consumer, True)

DTAnalytics.enable_log()

dt_id = "1234567890987654321"
acid = None

dt.user_unset("xx", "xxxx", {"#app_id": "aaa", "#bundle_id": "123", "bb": 2, "cc": "ss", "#is_foreground": True})
dt.user_delete("xx", "xxxx", {"#app_id": "aaa", "#bundle_id": "123", "bb": 2, "cc": "ss"})
dt.flush()

DTAnalytics.disable_log()

properties = dict()
properties["productNames"] = ["Lua", "hello"]
properties["productType"] = "Lua book"
properties["producePrice"] = 80
properties["shop"] = "xx-shop"
properties["#os"] = "1.1.1.1"
properties["sex"] = 'female'
properties["#app_id"] = "appid_1234567890"
properties["#bundle_id"] = "com.example"
for i in range(5):
    properties["a" + str(i)] = "asd" * i

n = 100000
tm = 0
lst = []
start_time = time.time()
for i in range(n):
    st = time.time()
    dt.track(dt_id, acid, "eventName", properties)
    crt = time.time() - st
    tm = tm + crt
    lst.append(crt)
end_time = time.time()
print('time elapsed: ' + str((end_time - start_time)*1000) + 'ms')
print('time elapsed avg: ' + str(tm / n * 1000) + 'ms')
lst.sort()
print('min: ' + str(lst[0]*1000) + "ms")
print("max: " + str(lst[-1]*1000) + "ms")
print("50': " + str(lst[int((len(lst)-1)*0.5)]*1000) + "ms")
print("80': " + str(lst[int((len(lst)-1)*0.8)]*1000) + "ms")
print("90': " + str(lst[int((len(lst)-1)*0.9)]*1000) + "ms")
print("95': " + str(lst[int((len(lst)-1)*0.95)]*1000) + "ms")
print("99': " + str(lst[int((len(lst)-1)*0.99)]*1000) + "ms")
num_write = int(n/200)
print(str((n-num_write)/n) + "': " + str(lst[-num_write-1]*1000) + "ms")
all_except_write = [lst[i] for i in range(len(lst)-num_write)]
print("avg (except write): " + str(sum(all_except_write) / len(all_except_write) * 1000) + "ms")
all_only_write = [lst[-1-i] for i in range(num_write)]
print("avg (write only): " + str(sum(all_only_write) / len(all_only_write) * 1000) + "ms")

dt.flush()
dt.close()

"""
Benchmark:
** 2024.04.09 **
QPS: 6500~7000
average: 0.146ms
average (except write): 0.030 ms
average (write only): 22.15ms
80': 0.032ms
90': 0.044ms
95': 0.055ms
99': 0.116ms
"""

File: dev/Benchmark.lua
-- $ lua dev/Benchmark.lua
socket = require("socket")

function getDivider()
    return package.config:sub(1,1)
end
function script_path()
    local str = debug.getinfo(2, "S").source:sub(2)
    local result = str:match("(.*" .. getDivider() .. ")")
    if result == nil then
        return ""
    end
    return result
end
local sdk_path = script_path() .. ".." .. getDivider() .. "lua" .. getDivider() .. "lua" .. getDivider()
package.path = package.path .. ";" .. sdk_path .. "?.lua"
local dtAnalytics = require("DataTowerSdk")

table.reduce = function (list, fn, init)
    local acc = init
    for k, v in ipairs(list) do
        if 1 == k and not init then
            acc = v
        else
            acc = fn(acc, v)
        end
    end
    return acc
end

table.sum = function(tb)
    return table.reduce(tb, function(a, b) return a + b end, 0)
end

local function getLogConsumer()
return dtAnalytics.DTLogConsumer("log", 200, 10 * 1024 * 1024)
end

dtAnalytics.enableLog(true)

local consumer = getLogConsumer()

--- init SDK with consumer
local sdk = dtAnalytics(consumer, false)

local dtId = "1234567890987654321"
local acId = nil

sdk:userSet("xx", "xxxx", { ["#app_id"] = "aaa", ["#bundle_id"] = "123", ["bb"] = 2, ["cc"] = "ss" })
sdk:flush()
dtAnalytics.enableLog(false)

local properties = {}
properties["productNames"] = { "Lua", "hello" }
properties["productType"] = "Lua book"
properties["producePrice"] = 80
properties["shop"] = "xx-shop"
properties["#os"] = "1.1.1.1"
properties["sex"] = 'female'
properties["#app_id"] = "appid_1234567890"
properties["#bundle_id"] = "com.example"
for i = 0, 5, 1
do
   properties["a" .. tostring(i)] = string.rep("asd", i)
end

n = 100000
tm = 0
local lst = {}
local start_time = socket.gettime()
for i = n, 1, -1
do
    properties["$_event_call_time"] = tostring(math.floor(socket.gettime() * 1000000))
    st = socket.gettime()
    sdk:track(acId, dtId, "eventName", properties)
    tmp = socket.gettime() - st
    tm = tm + tmp
    lst[#lst+1] = tmp
end
local end_time = socket.gettime()
print('time elapsed: ' .. ((end_time - start_time)*1000) .. 'ms')
print('time elapsed avg: ' .. (tm / n * 1000) .. 'ms')
table.sort(lst)
print('min: ' .. tostring(lst[1]*1000) .. "ms")
print("max: " .. tostring(lst[#lst]*1000) .. "ms")
print("50': " .. tostring(lst[math.floor((#lst-1)*0.5)]*1000) .. "ms")
print("80': " .. tostring(lst[math.floor((#lst-1)*0.8)]*1000) .. "ms")
print("90': " .. tostring(lst[math.floor((#lst-1)*0.9)]*1000) .. "ms")
print("95': " .. tostring(lst[math.floor((#lst-1)*0.95)]*1000) .. "ms")
print("99': " .. tostring(lst[math.floor((#lst-1)*0.99)]*1000) .. "ms")
num_write = math.floor(n/200)
print(tostring((n-num_write)/n) .. "': " .. tostring(lst[#lst-num_write]*1000) .. "ms")
all_except_write = {}
local i = 1
while (i <= #lst-num_write) do
    all_except_write[i] = lst[i]
    i = i + 1
end
print("avg (except write): " .. tostring(table.sum(all_except_write) / #all_except_write * 1000) .. "ms")
all_only_write = {}
local i2 = #lst-num_write+1
while (i2 <= #lst) do
    all_only_write[#all_only_write+1] = lst[i2]
    i2 = i2 + 1
end
print("avg (write only): " .. tostring(table.sum(all_only_write) / #all_only_write * 1000) .. "ms")

sdk:flush()
sdk:close()

--[[
Benchmark:
** 2024.04.10 **
QPS: 12000~15000
average: 0.0745ms
average (except write): 0.053 ms
average (write only): 4.256ms
80': 0.068ms
90': 0.083ms
95': 0.099ms
99': 0.160ms
]]--